Migration of Product Collection and Product Button blocks to the new `store()` API

luisherranz opened the pull request:
<!-- Please do not remove any information from this pull request. Instead, add N/A or leave blank if not applicable -->

## What

Migrates both the Product Collection and the Product Button block to the new store() API.

Resources:

- https://github.com/woocommerce/woocommerce-blocks/issues/11065
- https://github.com/woocommerce/woocommerce-blocks/pull/11071
- https://github.com/WordPress/gutenberg/discussions/53586

## Why

<!-- Describe the reason for your changes. This will help the reviewer and future readers get additional context -->

Because we are switching from the old `store()` API to the new `store()` API, and we need to migrate all the blocks at once.

## Testing Instructions

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Load a template that contains with Product Button blocks
2. Test that it keeps working as exactly as it is working on `trunk`

## Screenshots or screencast

I made a video to explain the migration of this block:

<div>
    <a href="https://www.loom.com/share/cbe30d0c820e4f99a3e6697a5f85b759">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/cbe30d0c820e4f99a3e6697a5f85b759-with-play.gif">
    </a>
  </div>

https://www.loom.com/share/cbe30d0c820e4f99a3e6697a5f85b759

## WooCommerce Visibility

<!-- Check this documentation link (../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WooCommerce core, part of the feature plugin, or experimental. -->
Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [ ] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog
<!-- Provide a brief, descriptive summary of the changes in this PR. Include potential impacts on different parts of the product. Example: "Updated the checkout process to streamline the experience for users and reduce the number of steps." -->

> Add suggested changelog entry here.

Migrate the Product Button to the new `store()` API of the Interactivity API.

luisherranz committed:
Refactor Product Button with new store() API

luisherranz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Use `wc_initial_state` in Product Button

DAreRodz committed:
Fix namespace

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Remove unnecessary state

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Test namespaces in directive paths

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Add test context with namespace

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Simplify woo-test context

DAreRodz committed:
Move addToCart and animations to a file

DAreRodz committed:
Do not pass `rawStore` to `afterLoad` callbacks

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

luisherranz committed:
Move callbacks and actions back to the main file

Because the animation was broken.

luisherranz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

luisherranz committed:
Remove selectors in favor of state

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Use default ns in `getContext` for state and actions

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Remove `afterLoad` callback

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Remove unnecessary ns

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Fix getContext in add-to-cart

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

luisherranz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

luisherranz committed:
Replace namespace and delete unnecessary store

luisherranz committed:
Pass context types only once

DAreRodz committed:
Add namespace to Product Collection block

luisherranz commented:
I think I'm going to leave the type of the async `addToCart` action as `() => void` instead of `() => Generator` for now, as it probably makes less sense to make TS think that that function returns a generator when it doesn't. Let's keep an eye on that, though.

I've documented this along with the need to cast getters in the original discussion:

- https://github.com/WordPress/gutenberg/discussions/53586#discussioncomment-7520599

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

DAreRodz committed:
Use an alternative for requestIdleCallback

DAreRodz committed:
Add previous react code for notices

DAreRodz commented:
The Product Collection block has required insignificant changes, so I included them in this PR instead of creating a new one.

I've updated the PR's title and description to reflect that. 🙂

imanish003 commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -64,237 +68,174 @@ const injectNotice = ( domNode: Element, errorMessage: string ) => {
-		return addToCartText;
-	}
-	return inTheCartText.replace( '###', numberOfItems.toString() );
+const getTextButton = (
```
Nitpicking: The function's name initially led me to expect it would yield a TextButton component, but it appears to actually return a string for display on a Button. Perhaps a rename to `getButtonText` or `getButtonTextBasedOnCartQuantity` could be more descriptive. However, please feel free to disregard this suggestion if you have a different perspective 🙂.

DAreRodz replied:
Makes total sense. I'll update it to `getButtonText`.

DAreRodz replied:
Fixed in https://github.com/woocommerce/woocommerce-blocks/commit/5938fef6a1411a6aab5957a24bacf201509b6e18.

imanish003 commented on src/BlockTypes/ProductCollection.php:
```diff
@@ -119,7 +119,7 @@ public function add_navigation_id_directive( $block_content, $block, $instance )
 					'wc-product-collection-' . $this->parsed_block['attrs']['queryId']
 				);
-				$p->set_attribute( 'data-wc-interactive', true );
+				$p->set_attribute( 'data-wc-interactive', wp_json_encode( array( 'namespace' => 'woocommerce' ) ) );
```
Just curious, We are using `woocommerce/product-button` namespace for Product button, but we aren't using `woocommerce/product-collection` for Product Collection block. Is it because we are just using client-side pagination, therefore, it doesn't have to do much with Store? 
Additionally, why can't we simply use true as a value? Does this relate to recent modifications in the Store API?

luisherranz replied:
> we aren't using woocommerce/product-collection for Product Collection block

Yeah, we're testing "block-based store parts". So let's switch to `woocommerce/product-collection` here as well.

_It's true that now the logic is abstracted in a `data-wc-link` directive, but we'll eventually replace it with custom store logic once you start adding things like loading indicators or accessibility, similar to what we did for the [Query block](https://github.com/WordPress/gutenberg/blob/c20462652e63ea3f9769ae8694ad8f48509f40b3/packages/block-library/src/query/view.js#L35-L78)._

> why can't we simply use true as a value? Does this relate to recent modifications in the Store API?

Yes, blocks now have a default namespace.

DAreRodz replied:
I'll change the namespace to `woocommerce/product-collection`. 🙂

DAreRodz replied:
Fixed in https://github.com/woocommerce/woocommerce-blocks/commit/cefa07cc6ea24245517df4c6fde56050e61efc79.

imanish003 reviewed (approved):
I've just finished testing the changes and I'm happy to report that everything is working perfectly! 🚀 Before we go ahead with merging the PR, it would be awesome to get a thumbs-up from either Luigi or Albert as well 🙂

By the way, a big shoutout to Luis for the attached video - it was super helpful and explained things brilliantly. Kudos to you! 🙌🏻

luisherranz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

Aljullu commented on assets/js/atomic/blocks/product-elements/button/frontend.tsx:
```diff
@@ -64,237 +68,174 @@ const injectNotice = ( domNode: Element, errorMessage: string ) => {
-				} );
+				return getTextButton(
+					context.addToCartText,
+					state.inTheCartText!,
```
Heads-up that there is an ESLint warning in this line and L121.

Aljullu reviewed (commented):
I really like how the code gets much simpler with these changes, good job @luisherranz!

The code changes look good to me and I did some smoke testing and everything is working as expected. However, I don't have much experience in how the Interactivity API worked before, so I would also like if @gigitux can take a look and give his approvall.

By the way, before merging there will need to be some changes to the PR itself: `WooCommerce Visibility` should be `WooCommerce Core` and we should add a `type` label (I guess `type: enhancement`?).

gigitux commented on src/BlockTypes/ProductButton.php:
```diff
@@ -168,20 +164,27 @@ protected function render( $attributes, $content, $block ) {
 
-			$div_directives = 'data-wc-context=\'' . wp_json_encode( $context, JSON_NUMERIC_CHECK ) . '\'';
+			$interactive = array(
+				'namespace' => 'woocommerce/product-button',
```
NIT: We could use using the variables to compose the namespace:

```suggestion
				'namespace' => $this->namespace . "/" . $this->block_name,
```

gigitux reviewed (changes requested):
Thanks for your precious work and the helpful video!

I just added a NIT, feel free to skip it! However, I noticed two things:

- A Circular dependency is detected:
<img width="878" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/4463174/4a0e4bcc-8fb7-4198-aa1e-ea44d85ee6ac">

- Currently, there is an E2E test that fails: "21 [blockTheme] › product-collection/product-collection.block_theme.spec.ts:29:2 › Product Collection › Renders product collection block correctly with 9 items (27.0s)". I already re-run it, but it failed again:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/d93ce2b6-6404-43e0-a52d-c2c3060bc599)

luisherranz commented:
> A Circular dependency is detected

We can [turn CHECK_CIRCULAR_DEPS off](https://github.com/woocommerce/woocommerce-blocks/blob/01015242dcd3b31da7d49f691062c4d3fe46e735/bin/webpack-configs.js#L40) in [the Interactivity API config](https://github.com/woocommerce/woocommerce-blocks/blob/01015242dcd3b31da7d49f691062c4d3fe46e735/bin/webpack-configs.js#L926).

gigitux commented:
> > A Circular dependency is detected
> 
> We can [turn CHECK_CIRCULAR_DEPS off](https://github.com/woocommerce/woocommerce-blocks/blob/01015242dcd3b31da7d49f691062c4d3fe46e735/bin/webpack-configs.js#L40) in [the Interactivity API config](https://github.com/woocommerce/woocommerce-blocks/blob/01015242dcd3b31da7d49f691062c4d3fe46e735/bin/webpack-configs.js#L926).

I didn't check the code, but circular dependency could introduce some issues that could be hard to find/debug. I would be more oriented to fix the circular dependency, but it is just an opinion based on my past experience.

luisherranz commented:
Oh, sorry, I didn't explain myself well enough. I wouldn't suggest changing how you manage your code. But that warning comes from the Interactivity API runtime, so it's ok to remove that circle dep check in the specific Webpack config of the Interactivity API runtime. It won't affect any Woo code.

gigitux commented:
> Oh, sorry, I didn't explain myself well enough. I wouldn't suggest changing how you manage your code. But that warning comes from the Interactivity API runtime, so it's ok to remove that circle dep check in the specific Webpack config of the Interactivity API runtime. It won't affect any Woo code.

Yeah, I got it! 
My suggestion was to fix it, even if it is in the Interactivity API runtime (not a blocker for this PR). But as I said, it is just an opinion that I have on circular dependency based on past experience :D

DAreRodz committed:
Replace getTextButton with getButtonText

DAreRodz committed:
Add block name to the ProductCollection namespace

luisherranz commented:
Thanks, Luigi. For now, we're trying to make things work and want to keep iterating fast. We'll make them _pretty_ after we move them to Gutenberg and we are about to merge them with WP Core 🙂

DAreRodz commented:
> By the way, before merging there will need to be some changes to the PR itself: `WooCommerce Visibility` should be `WooCommerce Core` and we should add a `type` label (I guess `type: enhancement`?).

I've updated the visibility and the type label. Feel free to modify them as needed. 🙂

BTW, this branch uses another PR as the base branch, but I added the changes just in case we decide to merge the base branch first to `trunk`.

gigitux committed:
fix style HTML code

gigitux commented:
With [1a3d9ca](https://github.com/woocommerce/woocommerce-blocks/pull/11558/commits/1a3d9ca50167a3af26ea49b6b20f56b27a80b914) I fixed the E2E test failure. For more details, check [this issue](https://github.com/woocommerce/woocommerce/issues/42315) (we should use `WP_HTML_Tag_Processor`). The other E2E tests fail due to the Product Gallery block and it is outside of the scope of this PR.

> Thanks, Luigi. For now, we're trying to make things work and want to keep iterating fast. We'll make them pretty after we move them to Gutenberg and we are about to merge them with WP Core

Thanks for the clarification. It makes sense 👍

gigitux reviewed (approved):
LGTM! Thanks for your work!

DAreRodz committed:
Merge branch 'interactivity-api-new-store-api' into interactivity-api-refactor-product-button-with-new-store-api

luisherranz committed:
Remove circular deps error on the Interactivity API

thealexandrelara committed:
Product Gallery block: Migrate to new Interactivity API store (#11721)

* Migrate Product Gallery block to new Interactivity API store

* Fix some references

* Add missing data-wc-interactive

* Fix an additional namespace

* Remove unnecessary click handler

* Dialog working

* Refactor action names

* Reindex PHP array

There was some missing indexes, which turned the array into an object in JS.

* Remove unused event handlers

* Move next/previous logic to external function

* Move StorePart util to the types folder

* Rename namespace to `woocommerce/product-gallery`

* Undo product collection namespace renaming

* Remove unnecessary namespace

* Don't hide the large image on page load

* Minor refactorings

* Fix eslint error

* Fix php cs errors with spacing and double arrows alignment

* Disable no-use-before-define rule for eslint

* Disable @typescript-eslint/ban-types rule for eslint

* Fix parsed context error in e2e tests

* Fix context parser for Thumbnail image

* Move store to the top of the frontend file

* Add interactivity api utils to the @woocommerce/utils alias

* Replace deprecated event attribute

---------

Co-authored-by: Luis Herranz <luisherranz@gmail.com>
Co-authored-by: David Arenas <david.arenas@automattic.com>
Co-authored-by: roykho <roykho77@gmail.com>

DAreRodz commented:
I've merged it even though some tests did not pass. 🤔 

To be honest, I'm wondering if they were failing because of a real issue in the code, because everything was green in https://github.com/woocommerce/woocommerce-blocks/pull/11721 and that branch already contained all the changes from this one (https://github.com/woocommerce/woocommerce-blocks/pull/11558), so they should be equivalent once it was merged here.
